### ISSUE TITLE
purchasePackage doesn't activate entitlement for returning users

### ISSUE THREAD

**Issue by 0x454447415244** [RevenueCat/react-native-purchases#357]
Hello,

The problem is that purchasePackage() does purchase the product but does not update active entitlements.

This problem happens when a user purchases the subscription product then deletes the app. If the user re-installs the app and attempts to purchase the product again, the active entitlement won't be updated.

For instance, after purchasing, the below if statement doesn't fire even if the purchase was successful:

```
const { purchaserInfo, productIdentifier } = await Purchases.purchasePackage(selectedProduct);

if (Object.entries(purchaserInfo.entitlements.active).length) { 
  //...
}
```

Please note that restoreTransactions() works. I'm using the latest stable version.

Thank you

**Comment by aboedo** [comment]
Hi! 👋 Thanks for opening. 

TL;DR: use `syncPurchases` to restore transactions after calling `configure` to achieve the behavior you're looking for. 


Longer explanation:
This is actually expected behavior - we don't proactively restore the transaction for the customer. The reason is that automatically restoring the purchase can lead to problematic behavior for developers who have an accounts system for an app, since when the app is reinstalled, the user won't be logged in, so restoring the transaction at that point might lead to unexpected behavior like the purchase being tied to the pre-login user, not the actual logged in user. 

The reason that the user can't re-purchase is that they already has an active subscription to that product, so they can't re-purchase so they don't get double-charged. 

A note about `restoreTransactions`: it works because the user already has the subscription to that product. But keep in mind that that method _could_ trigger a log-in prompt from the Store (App Store or Google Play) if the user's account isn't logged in. 

So because of this, and to prevent user confusion, we recommend only calling `restoreTransactions` after a user-initiated action, like a "Restore Purchases" button. 

`syncPurchases`, however, won't trigger a log in prompt, so I'd recommend using that one instead if you're looking to proactively restore purchases for a user when the app is opened.

More information about users, accounts and restores here: 
https://docs.revenuecat.com/docs/restoring-purchases
https://docs.revenuecat.com/docs/user-ids

**Comment by aboedo** [comment]
A couple of questions: 
- which running your app on iOS or Android?
- have you tried setting up [a listener for purchaser info updates](https://docs.revenuecat.com/docs/customer-info#listening-for-customerinfo-updates)?

**Comment by 0x454447415244** [comment]
I think I found the problem. I suspect that the id provided to the setup function is changing across installs.

If I don't provide an id, revenue cat will generate one. 

How does it generate the id? Is it based on the physical device?

**Comment by aboedo** [comment]
That makes sense. 
The random ID generated by RevenueCat is unique per physical device, and if you uninstall and reinstall, you'll also get a new random ID. 
More info here: https://docs.revenuecat.com/docs/user-ids#anonymous-app-user-ids

**Comment by 0x454447415244** [comment]
Shouldn't RevenueCat generate the same id for the same physical device? If not, I then see no way to fix the problem mentioned in this post (specially that my app does not register users).

**Comment by aboedo** [comment]
We don't generate the same ID for the same physical device, but that shouldn't be a problem for apps without user accounts systems, because anonymous IDs tied to the same Apple Receipt (which means the same Apple account for this purpose), will get aliased together and treated as one. 

For example, if a user installs the app, and a user `anonymousA` is generated, then they uninstall and reinstall, they'll become `anonymousB`. But if at this point `anonymousB` calls `restorePurchases` or `syncPurchases`, they will get the purchases made by `anonymousA`, because the receipt on the device is tied to the same Apple account. 

This is also useful for users who have multiple devices - for example, if a customer makes a purchase on an iPhone and then uses an iPad that has the same Apple account, the same exact flow applies - calling `restorePurchases` or `syncPurchases` will grant them access to the same purchases. 

I hope this explanation made sense! Let me know if you have any questions

**Comment by beylmk** [comment]
Closing this out since it seems the problem was addressed, but feel free to re-open if you're still experiencing issues or something is unclear!
